paper.tex

```diff
@@ -165,19 +179,25 @@ \subsection{Trading}
 	\label{fig:key_exchange}
 \end{figure}
 
+The whole trading process is shown in Figure \ref{fig:trading_product}. Once a consumer who want to subscribe the sensor data pays a subscription fee to the Product Contract, it is added to the consumers list automatically by the smart contract. To decrypt the streaming data on the MAM channel, session key $k$ should be exchanged between data provider and consumers as shown in Figure \ref{fig:key_exchange}.
+
+The data provider requests the consumers list maintained by the Product Contract and obtains accounts of each consumer. After that, the data provider can also get public keys of each consumer from the Registration Contract with their accounts. For each consumer, the data provider encrypts the session key and broker's signature with the consumer's public key and sends the ciphertext $Encrypt(k + Sign(k))$ to the Product Contract. Consumers listen to the smart contract event which is triggered when the ciphertext is updated, and decrypt the ciphertext to get the session key $k$ and signature $Sign(k)$. 
+
+Consumers can obtain the broker's public key on the Registration Contract as well, so they can verify that the signature is valid and the session key is the only one that is certified by the broker. Afterward, encrypted sensing data is uploaded to the MAM channel, and consumers can download and decrypt them with the session key.
+
 \subsection{Refunding}
-It is probable that the sensor is malfunctioning after the consumers pay the subscription fee. To protect consumers' right, the subscription fee are not transfered to the data provider until all data is generated and uploaded to the MAM channel. If the expected data is not available, consumers can ask for refunds. We assume that less than half of consumers in the Product Contract are irrational and malicious. Each consumer can vote for refunding at any time. If the ratio of consent votes of refunding is higher than the $threshold$ at $k$th piece of data, the subscription fee is proportionally transfered to the data provider, broker and every consumer. The subscription fee can be prorated as below:
+It is probable that the sensor is malfunctioning after the consumers pay the subscription fee. To protect consumers' right, the subscription fee are not transfered to the data provider until all data is generated and uploaded to the MAM channel. If the expected data is not available, consumers can ask for refunds. We assume that less than half of consumers in the Product Contract are irrational and malicious. Each consumer can vote for refunding at any time. If the ratio of consent votes of refunding is higher than the $threshold$ at the $i$th piece of data, the subscription fee is proportionally transfered to the data provider, broker and every consumer. The subscription fee can be prorated as below:
```

Comment:
Rather than referring to "sensor", how about changing it to general terminologies such as data source? Not all data was generated via sensors. Logging itself is an example of valuable data which does not depend on sensors.